include/triton/Dialect/Triton/IR/TritonOps.td

```diff
@@ -253,34 +271,57 @@ def TT_LoadOp : TT_Op<"load",
                        "triton::EvictionPolicy":$evict, "bool":$isVolatile)>,
         // A utility function to build the operation with all attributes
         OpBuilder<(ins "Value":$ptr, "Value":$mask, "Value":$other,
-                       "std::optional<ArrayRef<int32_t>>":$boundaryCheck,
+                       "ArrayRef<int32_t>":$boundaryCheck,
                        "std::optional<triton::PaddingOption>":$padding, "triton::CacheModifier":$cache,
                        "triton::EvictionPolicy":$evict, "bool":$isVolatile)>
     ];
 
-    // Format: `tt.load operands attrs : optional(type(ptr)) -> type(result)`
-    // We need an extra `optional(type(ptr))` for inferring the tensor pointer type with back compatibility
-    let hasCustomAssemblyFormat = 1;
+    // Specify `cacheModifier` and `evictionPolicy` explicitly in the
+    // assemblyFormat instead of as part of attr-dict so that they get printed
+    // as strings rather than opaque integers.
+    //
+    // Note there's no comma between `other` and `cacheModifier` and between
+    // `cacheModifier` and `evictionPolicy`.  This is due to an apparent
+    // limitation in the MLIR custom-format parser.  In oilist, the initial
+    // keywords of each clause have to be unique, so they can't be `,`.
+    //
+    // Even if we gave up on order-independence and used vanilla optional
+    // clauses, the format (`,` `foo` `=` $foo^)? (`,` `bar` `=` $bar^)?  will
+    // not match the string ", bar = 0" because after the initial comma (first
+    // token of the first optional clause) we expect to see "foo".
+    let assemblyFormat = [{
+      $ptr (`,` $mask^)? (`,` $other^)?
+      oilist(
+        `cacheModifier` `=` $cache |
+        `evictionPolicy` `=` $evict
+      )
+      attr-dict `:` type($ptr)
+    }];
 
     let hasCanonicalizer = 1;
 }
 
-def TT_StoreOp : TT_Op<"store",
-                       [SameLoadStoreOperandsShape,
-                        SameLoadStoreOperandsEncoding,
-                        MemoryEffects<[MemWrite<GlobalMemory>]>,
-                        TypesMatchWith<"infer ptr type from value type",
-                                       "value", "ptr", "$_self",
-                                       "mlir::OpTrait::impl::verifyLoadStorePointerAndValueType">,
-                        TypesMatchWith<"infer mask type from value type",
-                                       "value", "mask", "getI1SameShape($_self)",
-                                       "($_op.getOperands().size() <= 2) || std::equal_to<>()">]> {
+def TT_StoreOp : TT_Op<"store", [
+  SameLoadStoreOperandsShape,
+  SameLoadStoreOperandsEncoding,
+  MemoryEffects<[MemWrite<GlobalMemory>]>,
+  TypesMatchWith<"value type matches ptr type", "ptr", "value",
```

Comment:
This seems a breaking change for PT2. The torch inductor does epilog fusion which fuses one kernel into the other and relies on Triton to infer the pointer types here. cc @jansel

Reply:
IIUC, it now requires value type and pointer type to match and I'm seeing a breakage because of that.

Reply:
Despite its name, TypesMatchWith will do type inference.

If you doubt that, consider that the type of the mask also has to be inferred, and this happens with TypesMatchWith before and after this change.